netenv/location.go

```diff
@@ -91,72 +270,106 @@ next:
+				case layers.ICMPv4TypeEchoReply:
+					// Check if the ID and sequence match.
+					if icmpPacket.Id != uint16(msgID) {
+						continue listen
 					}
-					return nil, err
+					if icmpPacket.Seq < uint16(minSeq) {
+						continue listen
+					}
+					// We received a reply, so we did not trigger a time exceeded response on the way.
+					// This means we were not able to find the nearest router to us.
+					return false
+				case layers.ICMPv4TypeDestinationUnreachable,
+					layers.ICMPv4TypeTimeExceeded:
+					// Continue processing.
+				default:
+					continue listen
 				}
 
-				// Parse remote IP address.
-				addr, ok := src.(*net.IPAddr)
-				if !ok {
-					return nil, fmt.Errorf("failed to parse IP: %s", src.String())
+				// Parse copy of origin icmp packet that triggered the error.
+				if len(icmpPacket.Payload) != ipv4.HeaderLen+8 {
+					continue listen
 				}
-
-				// Continue if we receive a packet from ourself. This is specific to Windows.
-				if me, err := IsMyIP(addr.IP); err == nil && me {
-					log.Tracef("netenv: location approximation: ignoring own message from %s", src)
+				originalMessage, err := icmp.ParseMessage(1, icmpPacket.Payload[ipv4.HeaderLen:])
+				if err != nil {
 					continue listen
 				}
-
-				// If we received something from a global IP address, we have succeeded and can return immediately.
-				if netutils.GetIPScope(addr.IP).IsGlobal() {
-					return addr.IP, nil
+				originalEcho, ok := originalMessage.Body.(*icmp.Echo)
+				if !ok {
+					continue listen
 				}
-
-				// For everey non-global reply received, increase the maximum hops to try.
-				maxHops++
-
-				// Parse the ICMP message.
-				icmpReply, err := icmp.ParseMessage(1, recvBuffer[:n])
-				if err != nil {
-					log.Warningf("netenv: location approximation: failed to parse ICMP message: %s", err)
+				// Check if the ID and sequence match.
+				if originalEcho.ID != int(msgID) {
+					continue listen
+				}
+				if originalEcho.Seq < minSeq {
 					continue listen
 				}
 
 				// React based on message type.
-				switch icmpReply.Type {
-				case ipv4.ICMPTypeTimeExceeded, ipv4.ICMPTypeEchoReply:
-					log.Tracef("netenv: location approximation: receveived %q from %s", icmpReply.Type, addr.IP)
-					continue next
-				case ipv4.ICMPTypeDestinationUnreachable:
-					return nil, fmt.Errorf("destination unreachable")
-				default:
-					log.Tracef("netenv: location approximation: unexpected ICMP reply: received %q from %s", icmpReply.Type, addr.IP)
+				switch icmpPacket.TypeCode.Type() {
+				case layers.ICMPv4TypeDestinationUnreachable:
+					// We have received a valid destination unreachable response, abort.
+					return false
+
+				case layers.ICMPv4TypeTimeExceeded:
+					// We have received a valid time exceeded error.
+					// If message came from a global unicast, us it!
+					if netutils.GetIPScope(remoteIP) == netutils.Global {
+						return SetInternetLocation(remoteIP, SourceTraceroute)
+					}
```

Comment:
That check is done in SetInternetLocation as well so we can just call it and abort if it returns true.